fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1544,34 +1544,83 @@ Future<Void> deleteTenant(Reference<DB> db, int64_t id) {
 	return Void();
 }
 
-ACTOR template <class Transaction>
-Future<std::vector<std::pair<TenantName, TenantMapEntry>>> listTenantsTransaction(Transaction tr,
-                                                                                  TenantNameRef begin,
-                                                                                  TenantNameRef end,
-                                                                                  int limit) {
-	tr->setOption(FDBTransactionOptions::RAW_ACCESS);
+template <class Transaction>
+Future<std::vector<std::pair<TenantName, int64_t>>> listTenantsTransaction(Transaction tr,
+                                                                           TenantName begin,
+                                                                           TenantName end,
+                                                                           int limit,
+                                                                           int offset = 0) {
+	auto future = ManagementClusterMetadata::tenantMetadata().tenantNameIndex.getRange(tr, begin, end, limit + offset);
+	return fmap(
+	    [offset](auto f) {
+		    std::vector<std::pair<TenantName, int64_t>>& results = f.results;
+		    results.erase(results.begin(), results.begin() + offset);
+		    return results;
+	    },
+	    future);
+}
 
-	state KeyBackedRangeResult<std::pair<TenantName, int64_t>> matchingTenants =
-	    wait(ManagementClusterMetadata::tenantMetadata().tenantNameIndex.getRange(tr, begin, end, limit));
+template <class DB>
+Future<std::vector<std::pair<TenantName, int64_t>>> listTenants(Reference<DB> db,
+                                                                TenantName begin,
+                                                                TenantName end,
+                                                                int limit,
+                                                                int offset = 0) {
+	return runTransaction(db, [=](Reference<typename DB::TransactionT> tr) {
+		tr->setOption(FDBTransactionOptions::LOCK_AWARE);
```

Comment:
```suggestion
		tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
```

Reply:
I use `FDBTransactionOptions::LOCK_AWARE` on purpose, because `FDBTransactionOptions::READ_LOCK_AWARE` will mark the transaction as read-only. But `runTransaction` will mark call `tr->commit` regardless of the read-only, which will cause `transaction_read_only` being thrown.
Another solution is to have a function `runTransactionReadOnly()` to replace `runTransaction` here.

Reply:
Oh interesting, that's fine then